### PR TITLE
NLP Example

### PR DESCRIPTION
**Pull request by cmvcordova**
Added first example with basic within-template functionality based on a huggingface transformers example off the official lightning docs.

Things to watch out for:

- Method is currently tethered to the GLUE benchmark, to be extended into broadly using HF transformers.
- Training got "slower" after incorporating the SLURM copy call. Expected, needs verification.

### CODE REVIEW

File: docs/examples/profiling.ipynb
Comment (lebrice) [inline docs/examples/profiling.ipynb:1]: Probably shouldn't be included in this PR (belongs in another PR just for the profiling example)

File: docs/getting_started/install.md
Comment (lebrice) [inline docs/getting_started/install.md:1]: I think those changes were probably added by mistake. Better would be to keep the instructions from the `master` branch.

File: docs/examples/profiling.ipynb
Comment (cmvcordova) [reply docs/examples/profiling.ipynb:1]: removed

File: project/datamodules/text/hf_text.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+from lightning import LightningDataModule
+from torch.utils.data import DataLoader
+from transformers import (
+    AutoTokenizer,
+)
+
+from project.utils.env_vars import REPO_ROOTDIR, SCRATCH, SLURM_TMPDIR
+from project.utils.utils import get_task_info
+
+torch.set_num_threads(
```
Comment (lebrice) [inline project/datamodules/text/hf_text.py:19]: Also: If we do this here, it will always be run when importing the project, not just when using this datamodule.
Did this help make the test run faster?

File: project/main.py [hunk 2]
```diff
@@ -109,7 +109,7 @@ def evaluation(experiment: Experiment) -> tuple[str, float | None, dict]:
         # We want to report the training error.
         metrics = {
             **experiment.trainer.logged_metrics,
-            **experiment.trainer.callback_metrics,
```
Comment (lebrice) [inline project/main.py:112]: Why this change?

File: project/utils/utils.py [hunk 3]
```diff
@@ -156,3 +156,45 @@ def print_config(
 
     # with open("config_tree.log", "w") as file:
     #     rich.print(tree, file=file)
+
+
+## HF datamodule specific
+
+
+def get_task_info(task_name):
```
Comment (lebrice) [inline project/utils/utils.py:164]: A type hint would be nice, as it will be used by the auto-schema script and show the possible values in the datamodule config (assuming you also use it to annotate the constructor argument of the datamodule).

This should probably be moved to the same file as the  HF datamodule.

```suggestion
SupportedTask = Literal["cola", "sst2", "mrpc", "qqp", "stsb", "mnli", "qnli", "rte", "wnli", "ax"]
def get_task_info(task_name: SupportedTask):
```

**Review by lebrice — CHANGES_REQUESTED** [review]
Nice! Minor comments.

File: docs/install.md
[the remark this answers is not in the document]
Comment (cmvcordova) [reply docs/install.md:1]: fixed by a126d94dc5fd265443caa7d0771ca08d1625e14e

File: .gitignore
[the remark this answers is not in the document]
Comment (cmvcordova) [reply .gitignore:1]: fixed by a126d94dc5fd265443caa7d0771ca08d1625e14e

File: project/algorithms/hf_example.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+from project.datamodules.text.hf_text import HFDataModule
+
+
+def pretrained_network(model_name_or_path: str | Path, **kwargs) -> PreTrainedModel:
+    config = AutoConfig.from_pretrained(model_name_or_path, **kwargs)
+    return AutoModelForSequenceClassification.from_pretrained(model_name_or_path, config=config)
+
+
+class HFExample(LightningModule):
+    """General purpose lightning module for interfacing with huggingface transformers."""
```
[the remark this answers is not in the document]
Comment (cmvcordova) [reply project/algorithms/hf_example.py:24]: fixed by 7c6d2daa1cfce45d1c6015857629e6c50577257d

File: project/algorithms/hf_example.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        input_ids: torch.Tensor,
+        token_type_ids: torch.Tensor,
+        attention_mask: torch.Tensor,
+        labels: torch.Tensor,
+    ):
+        return self.network(
+            input_ids, token_type_ids=token_type_ids, attention_mask=attention_mask, labels=labels
+        )
+
+    def model_step(self, batch: dict):
```
[the remark this answers is not in the document]
Comment (cmvcordova) [reply project/algorithms/hf_example.py]: fixed by 7c6d2daa1cfce45d1c6015857629e6c50577257d

File: project/algorithms/hf_example.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        logits = outputs.logits
+
+        if self.num_labels > 1:
+            preds = torch.argmax(logits, axis=1)
+        else:
+            preds = logits.squeeze()
+
+        return loss, preds, labels
+
+    def training_step(self, batch, batch_idx):
```
[the remark this answers is not in the document]
Comment (cmvcordova) [reply project/algorithms/hf_example.py]: fixed by 7c6d2daa1cfce45d1c6015857629e6c50577257d

File: project/algorithms/hf_example.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        loss, preds, labels = self.model_step(batch)
+        self.log("train/loss", loss, on_step=False, on_epoch=True, prog_bar=True)
+        return {"loss": loss, "preds": preds, "labels": labels}
+
+    def validation_step(self, batch, batch_idx, dataloader_idx=0):
```
[the remark this answers is not in the document]
Comment (cmvcordova) [reply project/algorithms/hf_example.py]: fixed by b2fe0ef993e7b50f07ab04f5e57f2489fcf518a4

File: project/algorithms/hf_example.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        val_loss, preds, labels = self.model_step(batch)
+        self.log("val/loss", val_loss, on_step=False, on_epoch=True, prog_bar=True)
+        return {"val/loss": val_loss, "preds": preds, "labels": labels}
+
+    def on_validation_epoch_end(self) -> None:
+        pass
```
[the remark this answers is not in the document]
Comment (cmvcordova) [reply project/algorithms/hf_example.py]: fixed by b2fe0ef993e7b50f07ab04f5e57f2489fcf518a4

File: project/algorithms/testsuites/algorithm_tests.py [hunk 9]
```diff
@@ -179,16 +187,20 @@ def test_backward_pass_is_reproducible(
...
             )
         # BUG: Fix issue in tensor_regression calling .numpy() on cuda tensors.
-        assert (
-            isinstance(gradients_callback.batch, list | tuple)
-            and len(gradients_callback.batch) == 2
-        )
         assert isinstance(gradients_callback.grads, dict)
         assert isinstance(gradients_callback.outputs, dict)
+        batch = gradients_callback.batch
+        if isinstance(batch, list | tuple) and len(batch) == 2:
```
[the remark this answers is not in the document]
Comment (cmvcordova) [reply project/algorithms/testsuites/algorithm_tests.py]: fixed by 97b8b2af949612f12644f7c8a040e4d74065df8a

File: project/configs/datamodule/imagenet.yaml
[the remark this answers is not in the document]
Comment (cmvcordova) [reply project/configs/datamodule/imagenet.yaml:1]: fixed by 97b8b2af949612f12644f7c8a040e4d74065df8a

File: project/algorithms/hf_example_test.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+@run_for_all_configs_of_type("network", PreTrainedModel)
+class TestHFExample(LearningAlgorithmTests[HFExample]):
+    """Tests for the HF example."""
+
+    @pytest.fixture(scope="session")
+    def forward_pass_input(self, training_batch: dict[str, Tensor]):
+        assert isinstance(training_batch, dict)
+        return training_batch
+
+    @pytest.mark.skip(reason="Throughput problems")
```
[the remark this answers is not in the document]
Comment (cmvcordova) [reply project/algorithms/hf_example_test.py]: fixed by 97b8b2af949612f12644f7c8a040e4d74065df8a

File: project/datamodules/datamodules_test/overfit/__init__.py
[the remark this answers is not in the document]
Comment (cmvcordova) [reply project/datamodules/datamodules_test/overfit/__init__.py:1]: removed in 10bd60a25bd718b06128c4fe407f1dc4e37f2fbf

File: project/datamodules/datamodules_test/pretraining_location/__init__.py
[the remark this answers is not in the document]
Comment (cmvcordova) [reply project/datamodules/datamodules_test/pretraining_location/__init__.py:1]: removed in 10bd60a25bd718b06128c4fe407f1dc4e37f2fbf

File: project/datamodules/text/hf_text.py [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+from pathlib import Path
+
+import numpy as np
+import torch
+from datasets import DatasetDict, load_dataset
+from lightning import LightningDataModule
+from torch.utils.data import DataLoader
+from transformers import (
+    AutoTokenizer,
+)
```
[the remark this answers is not in the document]
Comment (cmvcordova) [reply project/datamodules/text/hf_text.py]: fixed in 568b537c2bd0543066cf3b9ec260fc778c8ceed3

File: project/datamodules/text/hf_text.py [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+
+from project.utils.env_vars import REPO_ROOTDIR, SCRATCH, SLURM_TMPDIR
+from project.utils.utils import get_task_info
+
+torch.set_num_threads(
+    1
+)  # https://github.com/Lightning-AI/pytorch-lightning/issues/10389#issuecomment-2310630247
```
[the remark this answers is not in the document]
Comment (cmvcordova) [reply project/datamodules/text/hf_text.py]: fixed in 568b537c2bd0543066cf3b9ec260fc778c8ceed3

File: project/datamodules/text/hf_text.py (hunk 1 above)
Comment (cmvcordova) [reply project/datamodules/text/hf_text.py]: fixed in 568b537c2bd0543066cf3b9ec260fc778c8ceed3

File: project/utils/utils.py (hunk 3 above)
Comment (cmvcordova) [reply project/utils/utils.py]: fixed by 1e989128562678b549fd48ac847119d0aad52311

File: project/main.py (hunk 2 above)
Comment (cmvcordova) [reply project/main.py]: fixed in b500fe5a1e47088f6112b207109a48e1bd4b49eb

File: project/algorithms/hf_example.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lebrice) [reply project/algorithms/hf_example.py:24]: Wrong commit perhaps? This didnt change.

File: project/algorithms/hf_example_test.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (lebrice) [reply project/algorithms/hf_example_test.py]: Do they eventually pass ? Or do they fail?

File: project/main.py (hunk 2 above)
Comment (lebrice) [reply project/main.py]: Watch out, this commit also reverted good changes in the top portion of the file!

File: project/utils/utils.py (hunk 3 above)
Comment (lebrice) [reply project/utils/utils.py]: Could also use the type in the constructor of the datamodule, so the auto_schema.py script will use it when contructing the schema for the config

File: project/datamodules/text/hf_text.py [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+
+class HFDataModule(LightningDataModule):  ## to be homogenized with the base text class
+    """Lightning data module for HF datasets."""
+
+    def __init__(
+        self,
+        hf_dataset_path: str,
+        tokenizer: str,
+        data_dir: str | Path = SCRATCH or REPO_ROOTDIR / "data",
+        task_name: str | None = None,
```
Comment (lebrice) [inline project/datamodules/text/hf_text.py:73]:
```suggestion
        task_name: SupportedTask | None = None,
```
Also, what happens if we pass `None`? Could we set a default explicitly here (or make it required) instead?

**Review by lebrice — APPROVED** [review]

File: project/main.py (hunk 2 above)
Comment (cmvcordova) [reply project/main.py]: addressed in be21fdbe98b4d1556ae63d6f22c03107992fba95

File: project/datamodules/text/hf_text.py (hunk 13 above)
Comment (cmvcordova) [reply project/datamodules/text/hf_text.py]: required in be21fdbe98b4d1556ae63d6f22c03107992fba95

File: project/datamodules/text/hf_text.py [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+class HFDataModule(LightningDataModule):  ## to be homogenized with the base text class
+    """Lightning data module for HF datasets."""
+
+    def __init__(
+        self,
+        hf_dataset_path: str,
+        tokenizer: str,
+        task_name: SupportedTask,
+        data_dir: str | Path = SCRATCH or REPO_ROOTDIR / "data",
+        task_field_map: dict | None = None,
```
Comment (lebrice) [inline project/datamodules/text/hf_text.py:74]:
```suggestion
```

File: project/datamodules/text/hf_text.py [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+            self.working_path = SLURM_TMPDIR / self.processed_dataset_path.name
+        else:
+            self.working_path = self.processed_dataset_path
+        # self.dataset_path = self.working_path = self.data_dir / f"{self.task_name}_dataset"
+
+        if task_name:
+            self.text_fields, self.num_labels = get_task_info(task_name)
+        else:
+            self.text_fields = task_field_map
+            self.num_labels = num_labels
```
Comment (lebrice) [inline project/datamodules/text/hf_text.py:124]:
```suggestion
        self.text_fields, self.num_labels = get_task_info(task_name)
```

File: project/datamodules/text/hf_text.py [hunk 16]
```diff
@@ -0,0 +1,277 @@
...
+    """Lightning data module for HF datasets."""
+
+    def __init__(
+        self,
+        hf_dataset_path: str,
+        tokenizer: str,
+        task_name: SupportedTask,
+        data_dir: str | Path = SCRATCH or REPO_ROOTDIR / "data",
+        task_field_map: dict | None = None,
+        num_labels: int | None = None,
```
Comment (lebrice) [inline project/datamodules/text/hf_text.py:75]:
```suggestion
```

File: project/datamodules/text/hf_text.py [hunk 17]
```diff
@@ -0,0 +1,277 @@
...
+
+        _rng = torch.Generator(device="cpu").manual_seed(self.seed)
+        self.train_dl_rng_seed = int(torch.randint(0, int(1e6), (1,), generator=_rng).item())
+        self.val_dl_rng_seed = int(torch.randint(0, int(1e6), (1,), generator=_rng).item())
+        self.test_dl_rng_seed = int(torch.randint(0, int(1e6), (1,), generator=_rng).item())
+
+    def prepare_data(self):
+        """
+        1.
+        """
```
Comment (lebrice) [inline project/datamodules/text/hf_text.py:141]:
```suggestion
```

File: project/datamodules/text/hf_text.py [hunk 18]
```diff
@@ -0,0 +1,277 @@
...
+        tokenized_dataset = dataset.map(
+            self.convert_to_features,
+            batched=True,
+            remove_columns=["label"],
+            load_from_cache_file=True,
+        )
+        logger.debug(f"Saving (overwriting) tokenized dataset at {self.processed_dataset_path}")
+        tokenized_dataset.save_to_disk(str(self.processed_dataset_path))
+
+        ## the SLURM path exists and it wasn't chosen as the DATA_DIR variable by default
```
Comment (lebrice) [inline project/datamodules/text/hf_text.py:159]:
```suggestion
```

**Review by lebrice — APPROVED** [review]
